Makefile.release

```diff
@@ -100,9 +100,37 @@ upload:
 
 .PHONY: docker-build
 docker-build: tar
+	# Steps:
+	# 1. Copy appropriate coredns binary to build/docker/linux/<arch>
+	# 2. Copy Dockerfile to build/docker/linux/<arch>
+	# 3. Replace base image from alpine:latest to <arch>/alpine:latest
+	# 4. Add COPY qemu-<arch>-static to Dockerfile
+	# 5. Add qemu-<arch>-static to build/docker/linux/<arch>
+	# 6. Set docker handler
+	# <arch>:
+	# arm: arm32v6 arm
+	# arm64: arm64v8 aarch64
+	rm -rf build/docker
 	for arch in $(LINUX_ARCH); do \
-	    tar -xzf release/$(NAME)_$(VERSION)_linux_$$arch.tgz ;\
-	    docker build -t coredns . ;\
+	    mkdir -p build/docker/linux/$$arch ;\
+	    tar -xzf release/$(NAME)_$(VERSION)_linux_$$arch.tgz -C build/docker/linux/$$arch ;\
+	    cp Dockerfile build/docker/linux/$$arch ;\
+	    if [ $$arch != amd64 ]; then \
+	        if [ $$arch == arm ]; then \
+	            BASEIMAGE=arm32v6 ;\
+	            QEMUARCH=arm ;\
+	        elif [ $$arch == arm64 ]; then \
+	            BASEIMAGE=arm64v8 ;\
+	            QEMUARCH=aarch64 ;\
+	        else \
+	            BASEIMAGE=$$arch ;\
+	            QEMUARCH=$$arch ;\
+	        fi ;\
+	        sed -e "s/alpine:latest/$$BASEIMAGE\\/alpine:latest\nCOPY qemu-$${QEMUARCH}-static \\/usr\\/bin\\//" -i build/docker/linux/$$arch/Dockerfile ;\
+	        docker run --rm --privileged multiarch/qemu-user-static:register --reset ;\
+	        curl -sSL --retry 5 https://github.com/multiarch/qemu-user-static/releases/download/v2.7.0/x86_64_qemu-$${QEMUARCH}-static.tar.gz | tar -xz -C build/docker/linux/$$arch ;\
```

Comment:
I don't understand this qemu curl? Why do we need this?

Reply:
In Dockerfile, each `RUN` command is like invoking `docker run`. That means creating process. In case the dev machine is `amd64`, and the base docker image is `arm`, then `arm` binary in docker image will not be able to run on your dev machine (`amd64`).

That is why qemu is needed to emulate arm processor on amd64 machines. 


Below is the partial output generated with `make -f Make.release docker-build`. 

If base image is `alpine:latest`, then apk tries to fetch `fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/x86_64/APKINDEX.tar.gz` (notice **x86_64**).

If base image is `arm32v6/alpine:latest`, then apk tries to fetch `http://dl-cdn.alpinelinux.org/alpine/v3.6/main/armhf/APKINDEX.tar.gz` (notice **armhf**)


Of course, if we could get rid of `RUN apk ...` step in Dockerfile, then we don't need qumu as `ADD coredns /coredns` does not invoke `docker run`.

```
Step 1/6 : FROM alpine:latest
 ---> 37eec16f1872
Step 2/6 : MAINTAINER Miek Gieben <miek@miek.nl> @miekg
 ---> Running in bbcc00312cc4
 ---> 36a4015c8c01
Removing intermediate container bbcc00312cc4
Step 3/6 : RUN set -x && apk --update add bind-tools ca-certificates openssl && update-ca-certificates && rm -rf /var/cache/apk/*
 ---> Running in b8fa5bff0082
+ apk --update add bind-tools ca-certificates openssl
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/community/x86_64/APKINDEX.tar.gz
(1/8) Installing libgcc (6.3.0-r4)
(2/8) Installing libxml2 (2.9.4-r4)
(3/8) Installing bind-libs (9.11.1_p1-r1)
(4/8) Installing bind-tools (9.11.1_p1-r1)
(5/8) Installing ca-certificates (20161130-r2)
(6/8) Installing libcrypto1.0 (1.0.2k-r0)
(7/8) Installing libssl1.0 (1.0.2k-r0)
(8/8) Installing openssl (1.0.2k-r0)
Executing busybox-1.26.2-r7.trigger
Executing ca-certificates-20161130-r2.trigger
OK: 13 MiB in 19 packages
+ update-ca-certificates
WARNING: ca-certificates.crt does not contain exactly one certificate or CRL: skipping
+ rm -rf /var/cache/apk/APKINDEX.24d64ab1.tar.gz /var/cache/apk/APKINDEX.84815163.tar.gz
 ---> 9ae2ecd5ebec
Removing intermediate container b8fa5bff0082
Step 4/6 : ADD coredns /coredns
 ---> e7f5e3e0fa3c
Removing intermediate container a362086352ac
Step 5/6 : EXPOSE 53 53/udp
 ---> Running in c1236c9db587
 ---> 6ffff3793d18
Removing intermediate container c1236c9db587
Step 6/6 : ENTRYPOINT /coredns
 ---> Running in 0b6756ba8ada
 ---> 881f507a4520
Removing intermediate container 0b6756ba8ada
Successfully built 881f507a4520
Setting /usr/bin/qemu-alpha-static as binfmt interpreter for alpha
Setting /usr/bin/qemu-arm-static as binfmt interpreter for arm
Setting /usr/bin/qemu-sparc32plus-static as binfmt interpreter for sparc32plus
Setting /usr/bin/qemu-ppc-static as binfmt interpreter for ppc
Setting /usr/bin/qemu-ppc64-static as binfmt interpreter for ppc64
Setting /usr/bin/qemu-ppc64le-static as binfmt interpreter for ppc64le
Setting /usr/bin/qemu-m68k-static as binfmt interpreter for m68k
Setting /usr/bin/qemu-mips-static as binfmt interpreter for mips
Setting /usr/bin/qemu-mipsel-static as binfmt interpreter for mipsel
Setting /usr/bin/qemu-mipsn32-static as binfmt interpreter for mipsn32
Setting /usr/bin/qemu-mipsn32el-static as binfmt interpreter for mipsn32el
Setting /usr/bin/qemu-mips64-static as binfmt interpreter for mips64
Setting /usr/bin/qemu-mips64el-static as binfmt interpreter for mips64el
Setting /usr/bin/qemu-sh4-static as binfmt interpreter for sh4
Setting /usr/bin/qemu-sh4eb-static as binfmt interpreter for sh4eb
Setting /usr/bin/qemu-s390x-static as binfmt interpreter for s390x
Setting /usr/bin/qemu-aarch64-static as binfmt interpreter for aarch64
Setting /usr/bin/qemu-hppa-static as binfmt interpreter for hppa
Sending build context to Docker daemon 29.88 MB
Step 1/7 : FROM arm32v6/alpine:latest
 ---> 8ecf96c9b538
Step 2/7 : COPY qemu-arm-static /usr/bin/
 ---> 0343e043bca8
Removing intermediate container 71c1dbb8a114
Step 3/7 : MAINTAINER Miek Gieben <miek@miek.nl> @miekg
 ---> Running in 0bcdcd8a91e3
 ---> a6856bd6e8a5
Removing intermediate container 0bcdcd8a91e3
Step 4/7 : RUN set -x && apk --update add bind-tools ca-certificates openssl && update-ca-certificates && rm -rf /var/cache/apk/*
 ---> Running in 8819cfc00bef
+ apk --update add bind-tools ca-certificates openssl
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/armhf/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/community/armhf/APKINDEX.tar.gz
(1/7) Installing libxml2 (2.9.4-r4)
(2/7) Installing bind-libs (9.11.1_p1-r1)
(3/7) Installing bind-tools (9.11.1_p1-r1)
(4/7) Installing ca-certificates (20161130-r2)
(5/7) Installing libcrypto1.0 (1.0.2k-r0)
(6/7) Installing libssl1.0 (1.0.2k-r0)
(7/7) Installing openssl (1.0.2k-r0)
Executing busybox-1.26.2-r5.trigger
Executing ca-certificates-20161130-r2.trigger
OK: 10 MiB in 18 packages
```